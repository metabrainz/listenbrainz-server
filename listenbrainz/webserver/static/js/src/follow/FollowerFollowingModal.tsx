import * as React from "react";
import { includes as _includes } from "lodash";

import Pill from "../components/Pill";
import FollowButton from "../FollowButton";
import APIService from "../APIService";

export type FollowerFollowingModalProps = {
  user: ListenBrainzUser;
};

type FollowerFollowingModalState = {
  activeMode: "follower" | "following";
  followerList: Array<ListenBrainzUser>;
  followingList: Array<ListenBrainzUser>;
};

export default class FollowerFollowingModal extends React.Component<
  FollowerFollowingModalProps,
  FollowerFollowingModalState
> {
  APIService: APIService;

  constructor(props: FollowerFollowingModalProps) {
    super(props);
    this.APIService = new APIService(`${window.location.origin}/1`);
    this.state = {
      activeMode: "follower",
      followerList: [],
      followingList: [],
    };

    // DO NOT CHANGE THIS ORDER
    // Otherwise react messes up the follow button props;
    // The way to fix this is to not pass the loggedInUserFollowsUser prop
    // into the FollowButton compoent.
    // TODO: fix this
    this.getFollowing();
    this.getFollowers();
  }

  getFollowers = () => {
    const { user } = this.props;
    this.APIService.getFollowersOfUser(user.name).then(
      ({ followers }: { followers: Array<{ musicbrainz_id: string }> }) => {
        this.setState({
          followerList: followers.map(({ musicbrainz_id }) => {
            return {
              name: musicbrainz_id,
            };
          }),
        });
      }
    );
  };

  getFollowing = () => {
    const { user } = this.props;
    this.APIService.getFollowingForUser(user.name).then(
      ({ following }: { following: Array<{ musicbrainz_id: string }> }) => {
        this.setState({
          followingList: following.map(({ musicbrainz_id }) => {
            return { name: musicbrainz_id };
          }),
        });
      }
    );
  };

  updateMode = (mode: "follower" | "following") => {
    this.setState({ activeMode: mode }, () => {
      const { activeMode } = this.state;
      if (activeMode === "follower") this.getFollowers();
      else this.getFollowing();
    });
  };

  loggedInUserFollowsUser = (user: ListenBrainzUser): boolean => {
    const { followingList } = this.state;
    return _includes(
      followingList.map((listEntry: ListenBrainzUser) => listEntry.name),
      user.name
    );
  };

  render() {
    const { user } = this.props;
    const { activeMode, followerList, followingList } = this.state;
    const activeModeList =
      activeMode === "follower" ? followerList : followingList;
    return (
      <>
        <div className="text-center follower-following-pills">
          <div className="btn-group btn-group-justified" role="group">
            <Pill
              active={activeMode === "follower"}
              type="secondary"
              onClick={() => this.updateMode("follower")}
            >
              Followers ({followerList.length})
            </Pill>
            <Pill
              active={activeMode === "following"}
              type="secondary"
              onClick={() => this.updateMode("following")}
            >
              Following ({followingList.length})
            </Pill>
          </div>
        </div>
        <div className="follower-following-list">
          {activeModeList.map((listEntry: ListenBrainzUser) => {
            return (
              <>
                <div key={listEntry.name}>
                  <a
                    href={`/user/${listEntry.name}`}
                    target="_blank"
                    rel="noopener noreferrer"
                  >
                    {listEntry.name}
                  </a>
                  <FollowButton
                    type="block"
                    user={{ name: listEntry.name }}
                    loggedInUser={user}
                    loggedInUserFollowsUser={this.loggedInUserFollowsUser(
                      listEntry
                    )}
                  />
                </div>
              </>
            );
          })}
        </div>
      </>
    );
  }
}
