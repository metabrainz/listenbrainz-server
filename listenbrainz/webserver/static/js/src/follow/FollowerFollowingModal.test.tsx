/*
 * listenbrainz-server - Server for the ListenBrainz project.
 *
 * Copyright (C) 2020 Param Singh <iliekcomputers@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

import * as React from "react";
import { shallow } from "enzyme";
import FollowerFollowingModal from "./FollowerFollowingModal";

const props = {
  user: { name: "foobar" },
  apiUrl: "http://0.0.0.0",
  loggedInUser: null,
  followerList: ["foo"],
  followingList: ["bar"],
  loggedInUserFollowsUser: () => true,
  updateFollowingList: () => {},
};

describe("<FollowerFollowingModal />", () => {
  it("renders", () => {
    const wrapper = shallow(<FollowerFollowingModal {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("updateMode()", () => {
  it("updates the mode correctly", () => {
    const wrapper = shallow<FollowerFollowingModal>(
      <FollowerFollowingModal {...props} />
    );
    const instance = wrapper.instance();

    // initial state after first fetch
    expect(instance.state.activeMode).toEqual("follower");

    // does nothing if the same mode as the current mode is passed
    instance.updateMode("follower");
    expect(instance.state.activeMode).toEqual("follower");

    // updates the mode correctly
    instance.updateMode("following");
    expect(instance.state.activeMode).toEqual("following");
  });
});
