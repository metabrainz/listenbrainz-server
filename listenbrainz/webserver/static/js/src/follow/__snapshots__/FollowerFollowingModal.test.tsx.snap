// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FollowerFollowingModal /> renders 1`] = `"<div class=\\"text-center follower-following-pills\\"><div class=\\"btn-group btn-group-justified\\" role=\\"group\\"><button type=\\"button\\" style=\\"background:#353070;border:2px solid #353070;font-weight:700;color:white;border-radius:24px;outline:none;padding:3px 12px;margin:2px 6px 12px 6px;box-sizing:border-box\\">Followers (1)</button><button type=\\"button\\" style=\\"background:transparent;border:2px solid #BBBBBB;color:#BBBBBB;border-radius:24px;outline:none;padding:3px 12px;margin:2px 6px 12px 6px;box-sizing:border-box\\">Following (1)</button></div></div><div class=\\"follower-following-list\\"><div><div><a href=\\"/user/foo\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">foo</a></div></div></div>"`;
